onnxruntime/hosting/http/predict_request_handler.cc

```diff
@@ -0,0 +1,133 @@
+    (context).response.insert("x-ms-request-id", (ms_request_id));                                    \
+    if (!client_request_id.empty()) {                                                                 \
+      (context).response.insert("x-ms-client-request-id", (client_request_id));                       \
+    }                                                                                                 \
+    auto json_error_message = CreateJsonError(http_error_code, (message));                            \
+    LOGS((*logger), VERBOSE) << json_error_message;                                                   \
+    (context).response.result(http_error_code);                                                       \
+    (context).response.body() = json_error_message;                                                   \
+    (context).response.set(http::field::content_type, "application/json");                            \
+  }
+
+static bool ParseRequestPayload(const HttpContext& context, SupportedContentType request_type,
+                                /* out */ PredictRequest& predictRequest, /* out */ http::status& error_code, /* out */ std::string& error_message);
+
+void Predict(const std::string& name,
+             const std::string& version,
+             const std::string& action,
+             /* in, out */ HttpContext& context,
+             HostingEnvironment& env) {
+  auto logger = env.GetLogger(context.request_id);
+  LOGS(*logger, VERBOSE) << "Name: " << name << " Version: " << version << " Action: " << action;
+
+  // We need to persist the "x-ms-client-request-id" field for the user to track the request from client side
+  std::string client_request_id{};
+  if (context.request.find("x-ms-client-request-id") != context.request.end()) {
+    client_request_id = context.request["x-ms-client-request-id"].to_string();
+    LOGS(*logger, VERBOSE) << "x-ms-client-request-id: [" << client_request_id << "]";
+  }
+
+  // Request and Response content type information
+  SupportedContentType request_type = GetRequestContentType(context);
+  SupportedContentType response_type = GetResponseContentType(context);
+  if (response_type == SupportedContentType::Unknown) {
+    GenerateErrorResponse(logger, http::status::bad_request, "Unknown 'Accept' header field in the request", context, context.request_id, client_request_id);
+  }
+
+  // Deserialize the payload
+  auto body = context.request.body();
+  PredictRequest predict_request{};
+  http::status error_code;
+  std::string error_message;
+  bool parse_succeeded = ParseRequestPayload(context, request_type, predict_request, error_code, error_message);
+  if (!parse_succeeded) {
+    GenerateErrorResponse(logger, error_code, error_message, context, context.request_id, client_request_id);
+    return;
+  }
+
+  // Run Prediction
+  protobufutil::Status status;
+  Executor executor(std::move(env), context.request_id);
```

Comment:
I don't understand why the executor takes ownership of `env`. What happens to the next request? Is `env` gone then?

Reply:
This is a good question and I asked people on my team about this last week. So this is my understanding: `std::move` takes an lvalue that refers to an object and returns an rvalue that still refers to the same object. It's the same as using `static_cast<HostingEnvironment&&>`.

Reply:
After talking through it with Klein we think this should be changed BACK to a `shared_ptr`. It's going to be shared across multiple requests and we are running into some undefined behavior here.

@pranavsharma